### PR TITLE
Introduce specs

### PR DESCRIPTION
**Pull request by glung**
# Scope

- Backport specs from vanilla redux. 
- Export specs as an artifact. 

To use it on a store : 
- pull dependency 
- create a test class that extends `StoreTest` and implement abstract method `abstract <S> Store<S> createStore(Reducer<S> reducer, S state);` 

# Implementation details 

- Declared as a module 
- Dependencies : assertj, mockito, junit.
- Enforce java7

API changes : 
- add `Subscription.EMPTY`
- add `Store.INIT`

### CODE REVIEW

**Comment by glung** [comment]
Sooo after I did it, I kind of changed my mind.

My proposal is : 
- Separate `specs` module 
- Kotlin specs because : 
  - this is not a runtime dependency
  - It forces to bear in mind the interoperability with Kotlin when designing the API

Thoughts ?

File: src/test/java/redux/api/StoreTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import redux.api.helpers.Reducers;
+import redux.api.helpers.State;
+import redux.api.helpers.Todo;
+
+public abstract class StoreTest {
+
+    abstract <S> Store<S> createStore(Reducer<State> reducer, State state);
+
+    @Test
+    public void passes_the_initial_action_and_the_initial_state() {
```
Comment (pardom-zz) [inline src/test/java/redux/api/StoreTest.java]: Not a fan of underscores in Java method names.

**Comment by pardom-zz** [comment]
As a Kotlin advocate, I would love to write tests in Kotlin/Spek. However, do you think this barrier to entry will discourage people from contributing to the ecosystem?

File: src/test/java/redux/api/StoreTest.java (hunk 1 above)
Comment (glung) [reply src/test/java/redux/api/StoreTest.java]: I find it easier to read like that. 

But I will not fight for that, as I agree convention is to use camelcase.

**Comment by beyondeye** [comment]
A little bit off topic but what do you think about [kotlintest](https://github.com/kotlintest/kotlintest)? Have you tried it?

**Comment by glung** [comment]
@beyondeye I did not. I tried only jUnit and Speks. 

This is a proposal with jUnit. The advantage is that it should be easily extendable to provide the `createStore` method. 

Does it sound good to you ?

**Comment by glung** [comment]
Note : I am trying to test it on another repo but `compile 'com.github.jvm-redux.jvm-redux-api:api:specs-SNAPSHOT'` can't be resolved. : /

**Comment by beyondeye** [comment]
yeah probably junit is a more proper choice in this project context

**Comment by glung** [comment]
Sorry, I finished it only today - I have been a little busy. But here it is. 

@pardom I implemented it against your store here (if you want to take a look) https://github.com/pardom/redux-kotlin/pull/16

@beyondeye I am sorry I could not implement it against yours because it does not implement the API already. 

Let me know what you think you all !

**Review by pardom-zz — APPROVED** [review]
LGTM

**Pull request merged**